### ISSUE TITLE
multiples includes

### ISSUE THREAD

**Issue by FrankSt** [andrus/linkrest-emberjs#1]
In EmberJS the query parameters are key values, so if I try: include={"path":"articles","mapBy":"title"}&include=articles.body , I only see the last parameter. Is there a way to solve this in linkrest?

**Comment by andrus** [comment]
Not sure I understand the question. Could you explain in more detail what the problem is?

**Comment by FrankSt** [comment]
Suppose I want invoices and include the client and the address ( from 3 tables invoice-->client-->address)

I should have 127.0.0.1:8080/invoices?include=client&include=client.address

to obtain the JSON which include invoices with the relation client and the relation client.address

Problem is if I do that in Ember I get 127.0.0.1:8080/invoices?include=client.address only the last one because the key is 'include'.

I am currently looking if I can force the adaptor in Ember. 

Learned already a lot about linkrest and it works fine now.

Frank

**Comment by andrus** [comment]
I think I understand the description now. I don't know client-side Ember well enough, though it looks strange that *request* key would affect how the *response* is interpreted. 

Also my latest work on relationships handling will result in include limitations until sideloading is implemented, as the responses no longer return full related objects, only the IDs. I wonder if that's related to your problems?

Before:
```json
{
   "invoices" : [
       {
           "id": 1,
           "client": {
                 "id" : 51,
                 "address": { ... }
           }
       }
   ]
}
```

Now:
```json
{
   "invoices" : [
       {
           "id": 1,
           "client": 51
       }
   ]
}
```

**Comment by FrankSt** [comment]
No, that is not the problem, LinkRest JSON is fine, but if I look in other examples I saw:

let query = { include: 'loans,loans.article' };

but that doesn't work in LinkRest it should be:

let query = { include: 'loans', include: 'loans.article' };

but that is giving problems, because in javascript this is a hash with the same key.

so I would get: /friends?include=loans.article

I just want a clean solution, currently I am looking in Ember to bypass the adaptor so I can make a string and not an hash (in jquery.get() I can use both)

**Comment by andrus** [comment]
Ah cool. Actually we can support this syntax in the adapter:
```
include='loans,loans.article'
```
If you can solve it client-side for now , it would be great. And eventually we can include it in LR Ember adapter.  Pull requests for LR adapter are also welcome.

**Comment by FrankSt** [comment]
I have a solution in Ember, but it is not very nice, when constructing the query-param string I can split it.
If you tell me where to look in LR Ember, I can make the changes there, and so I will learn a bit more about LinkRest?

thx

**Comment by andrus** [comment]
[Inside the adapter](https://github.com/andrus/linkrest-emberjs/blob/master/linkrest-emberjs/src/main/java/org/objectstyle/linkrest/emberjs/EmberJSAdapter.java) you can redefine internal services. The service that you need to override is [ITreeProcessor](https://github.com/nhl/link-rest/tree/master/link-rest/src/main/java/com/nhl/link/rest/runtime/parser/tree). I just looked at it, subclassing IncludeExcludeProcessor implementation should be pretty easy actually. 

And then writing some unit tests for this (following the existing examples).

**Comment by FrankSt** [comment]
Sorry, this is to new for me. I have tried to subclass IncludeExcludeProcessor into EmberJS IncludeExcludeProcessor and in the adaptor:
binder.bind(IncludeExcludeProcessor.class).to(EmberJSIncludeExcludeProcessor.class);
but that doesn't go to my routine, also do I have to subclass IncludeWorker and Excludeworker because they are private?

**Comment by andrus** [comment]
You should use interface as the binding key:

```java
   binder.bind(ITreeProcessor.class).to(EmberJSIncludeExcludeProcessor.class);
```

 And hopefully you won't need to subclass IncludeWorker. It is already called with a list of includes. You just need to build that list out of a single key.

Hopefully not. 

Andrus

**Comment by FrankSt** [comment]
Ok, now I get the right process of my EmberJSIncludeExcludeProcessor, but how can I get the IncludeWorker?
When importing: import com.nhl.link.rest.runtime.parser.tree.IncludeWorker; I get not visible.
So I can't use it and I suppose it is not wise to subclass this?

**Comment by andrus** [comment]
I just committed a change making them public

**Comment by FrankSt** [comment]
It works for me now on my copy, but I think you will have a better solution...

**Comment by andrus** [comment]
still feel free to send a pull request via GitHub. I will either use that or write my own version.

**Comment by FrankSt** [comment]
i will try, but wait for your commit, because now if have subclassed IncludeWorker en ExcludeWorker

**Comment by andrus** [comment]
These classes are already public on master.

**Comment by FrankSt** [comment]
Yes and it works fine, but I have no experience with GIT to pull a request, so I have to read a bit..(again)
